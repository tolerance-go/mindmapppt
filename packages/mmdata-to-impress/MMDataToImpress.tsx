import { MMData } from '@mindmapppt/mm-data'
import { flattenDeep } from 'lodash'
import sanitizeHtml from 'sanitize-html'
import { getSectionMetas, SectionMeta } from './getSectionMetas'

export const MMDataToImpress = ({
   data,
   imgSrcPrefix = '',
}: {
   data: MMData
   imgSrcPrefix?: string
}) => {
   const renderSection = (item: SectionMeta & { key?: string }) => {
      const { mmdata, deepLevel, z, x, y, key } = item

      const hasImg = !!mmdata.imgUrls?.[0]

      return (
         <div
            key={key ?? mmdata.id}
            className='step'
            data-deep-level={deepLevel}
            data-z={z}
            data-x={x}
            data-y={y}
         >
            {mmdata.imgUrls?.[0] ? (
               <img
                  src={`${imgSrcPrefix}${mmdata.imgUrls?.[0]}`}
                  alt={`${mmdata.title}-img`}
                  style={{
                     height: 180,
                     margin: '0 auto 3rem auto',
                     background: 'transparent',
                  }}
               ></img>
            ) : null}
            {hasImg ? <h3>{mmdata.title}</h3> : <h2>{mmdata.title}</h2>}
            {mmdata.contentHTML ? (
               <div
                  dangerouslySetInnerHTML={{
                     __html: sanitizeHtml(mmdata.contentHTML),
                  }}
               ></div>
            ) : null}
         </div>
      )
   }

   /** 渲染树第一层级，作为瀑布的第一层 */
   const renderSections = (
      items?: SectionMeta[],
      level: number = 0,
   ): (React.ReactElement | React.ReactElement[])[] => {
      return (
         items?.map((item, index) => {
            if (!item.children?.length) {
               return renderSection(item)
            }
            return [
               renderSection(item),
               renderSections(item.children, level + 1),
               level > 0
                  ? renderSection({
                       ...item,
                       key: item.mmdata.id + '_tail', // 防止重复
                    })
                  : [],
            ] as React.ReactElement[]
         }) ?? []
      )
   }

   return (
      <>
         <div className='fallback-message'>
            <p>
               您的浏览器 <b>不支持某些必备的功能</b>，
               所以你会看到这条简化的提示信息
            </p>
            <p>
               为了获得最佳体验，请使用最新版本 <b>Chrome</b>，<b>Safari</b> 或{' '}
               <b>Firefox</b> 的浏览器.
            </p>
         </div>
         <div className='impress-wrapper'>
            <div
               id='impress'
               data-transition-duration='1000'
               data-width='1024'
               data-height='768'
               // data-max-scale='3'
               // data-min-scale='0'
               data-perspective='1000'
               // data-autoplay='7'
            >
               {flattenDeep(renderSections(getSectionMetas(data)))}
            </div>
            <div id='impress-toolbar'></div>
         </div>
      </>
   )
}
