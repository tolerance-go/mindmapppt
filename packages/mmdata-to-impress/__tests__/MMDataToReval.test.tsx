import renderer from 'react-test-renderer'
import { MMDataToImpress } from '../MMDataToImpress'

test('MMDataToImpress', async () => {
   expect(
      renderer
         .create(
            <MMDataToImpress
               data={[
                  {
                     children: [
                        {
                           children: [],
                           id: 'df33ea79-6532-44db-966a-e50fefdf7205',
                           title: '分支主题 1',
                        },
                        {
                           children: [
                              {
                                 children: [],
                                 id: '76dd5dc1-b063-4671-b320-264817cf08b0',
                                 title: '子主题 1',
                              },
                              {
                                 children: [],
                                 id: '7f89b01d-0ea6-41ac-b03f-1f6a288d4579',
                                 title: '子主题 2',
                              },
                           ],
                           id: 'e4f67763-0c0f-40bb-9622-47a9b2c0cdda',
                           title: '分支主题 3',
                        },
                        {
                           children: [],
                           id: 'f0846f61-7b67-40eb-932a-7f29d1d81174',
                           title: '分支主题 3',
                        },
                        {
                           children: [],
                           id: '4e26e5f9-b301-48cd-8769-7353e1bbd592',
                           title: '分支主题 4',
                        },
                     ],
                     id: 'b6fba32a5998f4471032b9a135',
                     title: '中心主题',
                  },
               ]}
            />,
         )
         .toJSON(),
   ).toMatchSnapshot()
})
