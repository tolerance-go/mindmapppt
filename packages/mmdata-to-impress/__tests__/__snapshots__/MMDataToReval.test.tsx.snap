// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MMDataToImpress 1`] = `
[
  <div
    className="fallback-message"
  >
    <p>
      您的浏览器 
      <b>
        不支持某些必备的功能
      </b>
      ， 所以你会看到这条简化的提示信息
    </p>
    <p>
      为了获得最佳体验，请使用最新版本 
      <b>
        Chrome
      </b>
      ，
      <b>
        Safari
      </b>
       或
       
      <b>
        Firefox
      </b>
       的浏览器.
    </p>
  </div>,
  <div
    className="impress-wrapper"
  >
    <div
      data-height="768"
      data-perspective="1000"
      data-transition-duration="1000"
      data-width="1024"
      id="impress"
    >
      <div
        className="step"
        data-deep-level={0}
        data-x={0}
        data-y={0}
        data-z={-0}
      >
        <h2>
          中心主题
        </h2>
      </div>
      <div
        className="step"
        data-deep-level={1}
        data-x={-2250}
        data-y={0}
        data-z={-3000}
      >
        <h2>
          分支主题 1
        </h2>
      </div>
      <div
        className="step"
        data-deep-level={1}
        data-x={-750}
        data-y={0}
        data-z={-3000}
      >
        <h2>
          分支主题 3
        </h2>
      </div>
      <div
        className="step"
        data-deep-level={2}
        data-x={-750}
        data-y={-750}
        data-z={-6000}
      >
        <h2>
          子主题 1
        </h2>
      </div>
      <div
        className="step"
        data-deep-level={2}
        data-x={-750}
        data-y={750}
        data-z={-6000}
      >
        <h2>
          子主题 2
        </h2>
      </div>
      <div
        className="step"
        data-deep-level={1}
        data-x={-750}
        data-y={0}
        data-z={-3000}
      >
        <h2>
          分支主题 3
        </h2>
      </div>
      <div
        className="step"
        data-deep-level={1}
        data-x={750}
        data-y={0}
        data-z={-3000}
      >
        <h2>
          分支主题 3
        </h2>
      </div>
      <div
        className="step"
        data-deep-level={1}
        data-x={2250}
        data-y={0}
        data-z={-3000}
      >
        <h2>
          分支主题 4
        </h2>
      </div>
    </div>
    <div
      id="impress-toolbar"
    />
  </div>,
]
`;
